File: src/main/java/me/miki/shindo/utils/ColorUtils.java
package me.miki.shindo.utils;

import net.minecraft.client.renderer.GlStateManager;

import java.awt.*;

public class ColorUtils {

    public static Color getRainbow(int index, double speed, int alpha) {
        int angle = (int) ((System.currentTimeMillis() / speed + index) % 360);
        float hue = angle / 360f;
        int color = Color.HSBtoRGB(hue, 1F, 1);
        Color c = new Color(color);
        return new Color(c.getRed(), c.getGreen(), c.getBlue(), alpha);
    }

    public static Color interpolateColors(int speed, int index, Color start, Color end) {

        int angle = (int) (((System.currentTimeMillis()) / speed + index) % 360);
        angle = (angle >= 180 ? 360 - angle : angle) * 2;

        return ColorUtils.interpolateColorHue(start, end, angle / 360f);
    }

    public static Color interpolateColor(Color from, Color to, double delta) {
        int red = MathUtils.interpolateInt(from.getRed(), to.getRed(), delta);
        int green = MathUtils.interpolateInt(from.getGreen(), to.getGreen(), delta);
        int blue = MathUtils.interpolateInt(from.getBlue(), to.getBlue(), delta);
        int alpha = MathUtils.interpolateInt(from.getAlpha(), to.getAlpha(), delta);
        return new Color(red, green, blue, alpha);
    }

    private static Color interpolateColorHue(Color color1, Color color2, float amount) {

        amount = Math.min(1, Math.max(0, amount));

        float[] color1HSB = Color.RGBtoHSB(color1.getRed(), color1.getGreen(), color1.getBlue(), null);
        float[] color2HSB = Color.RGBtoHSB(color2.getRed(), color2.getGreen(), color2.getBlue(), null);

        Color resultColor = Color.getHSBColor(MathUtils.interpolateFloat(color1HSB[0], color2HSB[0], amount), MathUtils.interpolateFloat(color1HSB[1], color2HSB[1], amount), MathUtils.interpolateFloat(color1HSB[2], color2HSB[2], amount));

        return new Color(resultColor.getRed(), resultColor.getGreen(), resultColor.getBlue(), MathUtils.interpolateInt(color1.getAlpha(), color2.getAlpha(), amount));
    }

    public static float getHue(Color color) {
        return rgbToHsb(color)[0];
    }

    public static float getSaturation(Color color) {
        return rgbToHsb(color)[1];
    }

    public static float getBrightness(Color color) {
        return rgbToHsb(color)[2];
    }

    private static float[] rgbToHsb(Color color) {

        float[] hsv = new float[3];

        Color.RGBtoHSB(color.getRed(), color.getGreen(), color.getBlue(), hsv);

        return hsv;
    }

    public static String removeColorCode(String text) {
        return text.replaceAll("\\u00a7" + "1", "").replaceAll("\\u00a7" + "2", "").replaceAll("\\u00a7" + "3", "")
                .replaceAll("\\u00a7" + "4", "").replaceAll("\\u00a7" + "5", "").replaceAll("\\u00a7" + "6", "")
                .replaceAll("\\u00a7" + "7", "").replaceAll("\\u00a7" + "8", "").replaceAll("\\u00a7" + "9", "")
                .replaceAll("\\u00a7" + "a", "").replaceAll("\\u00a7" + "b", "").replaceAll("\\u00a7" + "c", "")
                .replaceAll("\\u00a7" + "d", "").replaceAll("\\u00a7" + "e", "").replaceAll("\\u00a7" + "f", "")
                .replaceAll("\\u00a7" + "g", "").replaceAll("\\u00a7" + "k", "").replaceAll("\\u00a7" + "l", "")
                .replaceAll("\\u00a7" + "m", "").replaceAll("\\u00a7" + "n", "").replaceAll("\\u00a7" + "o", "")
                .replaceAll("\\u00a7" + "r", "");
    }

    public static void setColor(int color, float alpha) {

        float r = (float) (color >> 16 & 255) / 255.0F;
        float g = (float) (color >> 8 & 255) / 255.0F;
        float b = (float) (color & 255) / 255.0F;

        GlStateManager.color(r, g, b, alpha);
    }

    public static Color getColorByInt(int color) {

        float r = (float) (color >> 16 & 255) / 255.0F;
        float g = (float) (color >> 8 & 255) / 255.0F;
        float b = (float) (color & 255) / 255.0F;
        float a = (float) (color >> 24 & 255) / 255.0F;

        return new Color(r, g, b, a);
    }

    public static void setColor(int color) {
        setColor(color, (float) (color >> 24 & 255) / 255.0F);
    }

    public static void resetColor() {
        GlStateManager.color(1F, 1F, 1F, 1F);
    }

    public static Color applyAlpha(Color color, int alpha) {

        int r = color.getRed();
        int g = color.getGreen();
        int b = color.getBlue();

        return new Color(r, g, b, alpha);
    }

    public static Color lighten(Color color, float amount) {
        amount = Math.max(0F, Math.min(1F, amount));
        int r = color.getRed();
        int g = color.getGreen();
        int b = color.getBlue();
        int a = color.getAlpha();

        int newR = (int) Math.min(255, r + ((255 - r) * amount));
        int newG = (int) Math.min(255, g + ((255 - g) * amount));
        int newB = (int) Math.min(255, b + ((255 - b) * amount));

        return new Color(newR, newG, newB, a);
    }

    public static Color darken(Color color, float amount) {
        amount = Math.max(0F, Math.min(1F, amount));
        int r = color.getRed();
        int g = color.getGreen();
        int b = color.getBlue();
        int a = color.getAlpha();

        int newR = (int) Math.max(0, r - (r * amount));
        int newG = (int) Math.max(0, g - (g * amount));
        int newB = (int) Math.max(0, b - (b * amount));

        return new Color(newR, newG, newB, a);
    }

    public static float getAlphaByInt(int color) {

        return (float) (color >> 24 & 255) / 255.0F;
    }

    public static void color(int color) {
        float alpha = (color >> 24 & 255) / 255f;
        float red = (color >> 16 & 255) / 255f;
        float green = (color >> 8 & 255) / 255f;
        float blue = (color & 255) / 255f;
        GlStateManager.color(red, green, blue, alpha);
    }
}

File: src/main/kotlin/me/miki/shindo/utils/ColorUtils.kt
package me.miki.shindo.utils

import net.minecraft.client.renderer.GlStateManager
import java.awt.Color
import kotlin.math.max
import kotlin.math.min

object ColorUtils {

    @JvmStatic
    fun getRainbow(index: Int, speed: Double, alpha: Int): Color {
        val angle = ((System.currentTimeMillis() / speed + index) % 360).toInt()
        val hue = angle / 360f
        val color = Color.HSBtoRGB(hue, 1f, 1f)
        val base = Color(color)
        return Color(base.red, base.green, base.blue, alpha)
    }

    @JvmStatic
    fun interpolateColors(speed: Int, index: Int, start: Color, end: Color): Color {
        var angle = ((System.currentTimeMillis() / speed + index) % 360).toInt()
        angle = (if (angle >= 180) 360 - angle else angle) * 2
        return interpolateColorHue(start, end, angle / 360f)
    }

    @JvmStatic
    fun interpolateColor(from: Color, to: Color, delta: Double): Color {
        val red = MathUtils.interpolateInt(from.red, to.red, delta)
        val green = MathUtils.interpolateInt(from.green, to.green, delta)
        val blue = MathUtils.interpolateInt(from.blue, to.blue, delta)
        val alpha = MathUtils.interpolateInt(from.alpha, to.alpha, delta)
        return Color(red, green, blue, alpha)
    }

    private fun interpolateColorHue(color1: Color, color2: Color, amount: Float): Color {
        val amt = min(1f, max(0f, amount))

        val color1HSB = Color.RGBtoHSB(color1.red, color1.green, color1.blue, null)
        val color2HSB = Color.RGBtoHSB(color2.red, color2.green, color2.blue, null)

        val result = Color.getHSBColor(
            MathUtils.interpolateFloat(color1HSB[0], color2HSB[0], amt.toDouble()),
            MathUtils.interpolateFloat(color1HSB[1], color2HSB[1], amt.toDouble()),
            MathUtils.interpolateFloat(color1HSB[2], color2HSB[2], amt.toDouble())
        )

        return Color(
            result.red,
            result.green,
            result.blue,
            MathUtils.interpolateInt(color1.alpha, color2.alpha, amt.toDouble())
        )
    }

    @JvmStatic
    fun getHue(color: Color): Float = rgbToHsb(color)[0]

    @JvmStatic
    fun getSaturation(color: Color): Float = rgbToHsb(color)[1]

    @JvmStatic
    fun getBrightness(color: Color): Float = rgbToHsb(color)[2]

    private fun rgbToHsb(color: Color): FloatArray {
        val hsv = FloatArray(3)
        Color.RGBtoHSB(color.red, color.green, color.blue, hsv)
        return hsv
    }

    @JvmStatic
    fun removeColorCode(text: String): String {
        return text
            .replace("\\u00a71".toRegex(), "")
            .replace("\\u00a72".toRegex(), "")
            .replace("\\u00a73".toRegex(), "")
            .replace("\\u00a74".toRegex(), "")
            .replace("\\u00a75".toRegex(), "")
            .replace("\\u00a76".toRegex(), "")
            .replace("\\u00a77".toRegex(), "")
            .replace("\\u00a78".toRegex(), "")
            .replace("\\u00a79".toRegex(), "")
            .replace("\\u00a7a".toRegex(), "")
            .replace("\\u00a7b".toRegex(), "")
            .replace("\\u00a7c".toRegex(), "")
            .replace("\\u00a7d".toRegex(), "")
            .replace("\\u00a7e".toRegex(), "")
            .replace("\\u00a7f".toRegex(), "")
            .replace("\\u00a7g".toRegex(), "")
            .replace("\\u00a7k".toRegex(), "")
            .replace("\\u00a7l".toRegex(), "")
            .replace("\\u00a7m".toRegex(), "")
            .replace("\\u00a7n".toRegex(), "")
            .replace("\\u00a7o".toRegex(), "")
            .replace("\\u00a7r".toRegex(), "")
    }

    @JvmStatic
    fun setColor(color: Int, alpha: Float) {
        val r = (color shr 16 and 255) / 255.0f
        val g = (color shr 8 and 255) / 255.0f
        val b = (color and 255) / 255.0f
        GlStateManager.color(r, g, b, alpha)
    }

    @JvmStatic
    fun getColorByInt(color: Int): Color {
        val r = (color shr 16 and 255) / 255.0f
        val g = (color shr 8 and 255) / 255.0f
        val b = (color and 255) / 255.0f
        val a = (color shr 24 and 255) / 255.0f
        return Color(r, g, b, a)
    }

    @JvmStatic
    fun setColor(color: Int) {
        setColor(color, (color shr 24 and 255) / 255.0f)
    }

    @JvmStatic
    fun resetColor() {
        GlStateManager.color(1f, 1f, 1f, 1f)
    }

    @JvmStatic
    fun applyAlpha(color: Color, alpha: Int): Color {
        return Color(color.red, color.green, color.blue, alpha)
    }

    @JvmStatic
    fun lighten(color: Color, amount: Float): Color {
        val clamped = max(0f, min(1f, amount))
        val r = color.red
        val g = color.green
        val b = color.blue
        val a = color.alpha

        val newR = min(255f, r + ((255 - r) * clamped)).toInt()
        val newG = min(255f, g + ((255 - g) * clamped)).toInt()
        val newB = min(255f, b + ((255 - b) * clamped)).toInt()

        return Color(newR, newG, newB, a)
    }

    @JvmStatic
    fun darken(color: Color, amount: Float): Color {
        val clamped = max(0f, min(1f, amount))
        val r = color.red
        val g = color.green
        val b = color.blue
        val a = color.alpha

        val newR = max(0f, r - (r * clamped)).toInt()
        val newG = max(0f, g - (g * clamped)).toInt()
        val newB = max(0f, b - (b * clamped)).toInt()

        return Color(newR, newG, newB, a)
    }

    @JvmStatic
    fun getAlphaByInt(color: Int): Float {
        return (color shr 24 and 255) / 255.0f
    }

    @JvmStatic
    fun color(color: Int) {
        val alpha = (color shr 24 and 255) / 255f
        val red = (color shr 16 and 255) / 255f
        val green = (color shr 8 and 255) / 255f
        val blue = (color and 255) / 255f
        GlStateManager.color(red, green, blue, alpha)
    }
}
